### PR TITLE
copr: split copr requests by buckets

### LINKED ISSUE #32968: split copr requests by buckets
## Enhancement

Now TiKV supports 10GB size region, so copr requests should be splitted in buckets level for concurrency.

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve?

Issue Number: close #32968

Problem Summary:

### What is changed and how it works?

- Update client-go(https://github.com/tikv/client-go/pull/439) and pd-client(https://github.com/tikv/pd/pull/4699).
- Split copr requests by buckets
- If the `latestBucketsVersion` in copr reponse is newer than the cached one, update buckets asynchronously.
- No behaviour change if bucket is not enabled.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

Documentation

### Release note

```release-note
None
```

### CODE REVIEW

File: store/copr/coprocessor.go [hunk 1]
```diff
@@ -180,7 +181,8 @@ func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *KeyRanges, req *kv
 
 	rangesLen := ranges.Len()
 
-	locs, err := cache.SplitKeyRangesByLocations(bo, ranges)
+	// TODO(youjiali1995): is there any request type that needn't be splitted by buckets?
```
Comment (cfzjywxk) [inline store/copr/coprocessor.go:184]: Will the split affect some internal work such as br cheksum, data analyze or statistic update?

File: store/copr/coprocessor_test.go [hunk 2]
```diff
@@ -53,105 +53,252 @@ func TestBuildTasks(t *testing.T) {
...
 	require.NoError(t, err)
 	require.Len(t, tasks, 2)
-	taskEqual(t, tasks[0], regionIDs[1], "h", "k", "m", "n")
-	taskEqual(t, tasks[1], regionIDs[2], "n", "p")
+	taskEqual(t, tasks[0], regionIDs[1], 0, "h", "k", "m", "n")
+	taskEqual(t, tasks[1], regionIDs[2], 0, "n", "p")
+}
+
+func TestBuildTasksByBuckets(t *testing.T) {
+	mockClient, cluster, pdClient, err := testutils.NewMockTiKV("", nil)
```
Comment (cfzjywxk) [inline store/copr/coprocessor_test.go:157]: Should we use the new mockstore(`NewMockStore`) instead?

File: store/copr/coprocessor.go (hunk 1 above)
Comment (youjiali1995) [reply store/copr/coprocessor.go:184]: Yes, they should be splitted by buckets too.

File: store/copr/coprocessor_test.go (hunk 2 above)
Comment (youjiali1995) [reply store/copr/coprocessor_test.go:157]: No, `NewMockStore` doesn't export region schedule functions.

**Review by sticnarf — COMMENTED** [review]
`SplitKeyRangesByLocations` and `splitKeyRangesByBuckets` share the similar structure because a `bucket` is just like a region in region. Is it possible to unify them?

File: store/copr/coprocessor.go (hunk 1 above)
Comment (youjiali1995) [reply store/copr/coprocessor.go:184]: Yes. TiFlash has implemented server-side streaming by `BatchCoprocessor` which splits ranges by segments.

File: store/copr/coprocessor.go [hunk 3]
```diff
@@ -319,7 +322,6 @@ type copIteratorWorker struct {
 
 	replicaReadSeed uint32
 
-	actionOnExceed             *rateLimitAction
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply store/copr/coprocessor.go:322]: I don't know, but it's not used now.

File: store/copr/coprocessor.go [hunk 4]
```diff
@@ -146,8 +146,9 @@ func (c *CopClient) Send(ctx context.Context, req *kv.Request, variables interfa
 
 // copTask contains a related Region and KeyRange for a kv.Request.
 type copTask struct {
-	region tikv.RegionVerID
-	ranges *KeyRanges
+	region     tikv.RegionVerID
+	bucketsVer uint64
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply store/copr/coprocessor.go:150]: I forget it... Now it's used for reducing UpdateBucketsIfNeeded https://github.com/pingcap/tidb/pull/32818/commits/0b24a84e315b32a082c982ca9f8fe870dd795451#diff-23eb1043649369d5956b123b173b799ded8c6b7cbf4519bb32cb3caa01928d75R935.

**Review by sticnarf — APPROVED** [review]
LGTM

File: store/copr/region_cache.go [hunk 5]
```diff
@@ -63,6 +63,75 @@ type LocationKeyRanges struct {
...
+				}
+			}
+			res = append(res, &LocationKeyRanges{l.Location, ranges.Slice(0, i)})
+			ranges = ranges.Slice(i, ranges.Len())
+		} else {
+			// Iterate to the first range that is not complete in the bucket.
+			var i int
+			for ; i < ranges.Len(); i++ {
+				r := ranges.At(i)
+				if !(bucket.Contains(r.EndKey) || bytes.Equal(bucket.EndKey, r.EndKey)) {
```
[the remark this answers is not in the document]
Comment (youjiali1995) [reply store/copr/region_cache.go:99]: Should be `len(bucket.EndKey) != 0 && bytes.Compare(r.EndKey, bucket.EndKey) > 0`. It's not simpler.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Pull request merged**